drivers: counter: add Maxim DS3231 support

pabigot opened the pull request:
This PR provides support for the [Maxim DS3231 TCXO RTC](https://www.maximintegrated.com/en/products/analog/real-time-clocks/DS3231.html), a high-precision real-time clock peripheral with support for battery backup.

This is implemented as a counter, but the DS3231 has significant functionality enhancements beyond a general RTC API.    Also, as an I2C peripheral it has limitations that are incompatible with assumptions about how a counter can function.  Most users of its advanced functionality will interact with it through device specific API provided in a driver-specific header following the approach summarized in #17072.  Consequently the interface file is placed in the (existing) `include/drivers/rtc` directory, and the extension functions are named based on it being a specialization of a (no longer existing) `rtc` driver class.

There is a question of where to collect the documentation.  Because it's device specific, it really should be its own page at API Reference : Peripherals : RTC : DS3231.  For now I've just put it under (newly created) RTC.

As more infrastructure is added to support management of civil time tracking in Zephyr some of the functionality in this driver may need to be generalized. For example, the `rtc_ds3231_syncpoint` structure should probably be generalized and placed in the recently created `<sys/timeutil.h>` header long with functions that support generic conversion from a synchronization point and local time to RTC and back.

pabigot committed:
drivers: counter: add Maxim DS3231 support

The DS3231 is an I2C real-time clock with internal temperature
compensated oscillator, maintaining civil time to 1 s precision with
nominal 2 ppm accuracy from 0-40 Cel.

The basic functionality is exposed as a counter that is always running
at 1 Hz.  Much more functionality is exposed as driver-specific API,
including the ability to translate between the time scale of the DS3231
and the time scale of the Zephyr uptime clock.  This allows correlation
of events in the system clock to UTC, TAI, or whatever time scale is
used to maintain the DS3231.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

dbkinder commented on doc/reference/peripherals/rtc.rst:
```diff
@@ -0,0 +1,13 @@
+
+Overview
+********
+
```
Missing an overview here...

pabigot replied:
Yeah, I just copied counter.  The only peripherals that have overview are I2C and sensors.

Part of the reason I can't add one is that this is for RTC peripherals in general, and the PR provides a device-specific API.  I need instruction on how to document that.  From the PR description:
> A final question is where to collect the documentation. Because it's device specific, it really should be its own page at API Reference : Peripherals : RTC : DS3231. For now I've just put it under (newly created) RTC.

I've applied the other changes; thanks.

dbkinder commented on samples/drivers/rtc_ds3231/README.rst:
```diff
@@ -0,0 +1,155 @@
+Annotated Example Output
+************************
+
+The sample first displays the boot banner; board identifier and
```
use two commas (``,``) in this sentence instead of two semicolons (``;``)

dbkinder commented on samples/drivers/rtc_ds3231/README.rst:
```diff
@@ -0,0 +1,155 @@
+
+   DS3231 ctrl 04 ; ctrl_stat 08
+
+Next, if the sample application option
```
If you write this like so, it will link to the config option documentation:
```
Next, if the sample application option :option:`CONFIG_APP_SET_ALIGNED_CLOCK`
is set, the civil time will be advanced to ...
```

dbkinder commented on samples/drivers/rtc_ds3231/README.rst:
```diff
@@ -0,0 +1,155 @@
+   Min alarm flags e at 252374400 ~ 1977-12-31 00:00:00 Sat 365
+
+The second-resolution alarm was signalled, and processed by the
+application at 7.002 s into run, as scheduled (plus callback latency).
```
into ``the`` run

dbkinder commented on samples/drivers/rtc_ds3231/README.rst:
```diff
@@ -0,0 +1,155 @@
+********************
+
+Wire a Chronodot to one of the supported boards as specified in
+corresponding devicetree overlay.
```
``the`` corresponding

dbkinder reviewed (approved):
Sigh.  OK for now.  I'll add checking for overviews for the API docs to the bin list :)

dbkinder reviewed (changes requested):
The doc build is failing because there's no visible option for CONFIG_APP_SET_ALIGNED_CLOCK defined in the Kconfig tree?

dbkinder commented on samples/drivers/rtc_ds3231/README.rst:
```diff
@@ -0,0 +1,155 @@
+   DS3231 ctrl 04 ; ctrl_stat 08
+
+Next, if the sample application option
+:option:`CONFIG_APP_SET_ALIGNED_CLOCK` is set, the civil time will be
```
Sorry, my suggestion to use :option:`CONFIG_APP_ALIGNED_CLOCK` doesn't work when the config option is a "private" config option for a sample (and not part of the big KConfig tree).  So this would need to be back to the way you had it as simply 
```
Next, if the sample application option ``CONFIG_APP_SET_ALIGNED_CLOCK`` is set ...
```

pabigot replied:
Reverted.

dbkinder reviewed (approved):
+1 for doc changes, thanks.

carlescufi commented:
@anangl could you please take a look at the proposed API extension mechanism?

pabigot commented:
Added to dev-review in hopes of getting feedback on this specifically, and device-specific APIs in general.  TSC mention and posting on devel didn't produce much/anything.  Base question: can this make it into 2.0.0?

nordic-krch commented on drivers/counter/ds3231.c:
```diff
@@ -0,0 +1,1060 @@
+		return -EBUSY;
+	}
+
+	time_t when = alarm_cfg->ticks;
```
this assignment is redundant in case of absolute. What about moving it to else clause (if not absolute)?

pabigot replied:
done

nordic-krch commented on drivers/counter/ds3231.c:
```diff
@@ -0,0 +1,1060 @@
+
+	data->counter_handler[id] = NULL;
+	data->counter_ticks[id] = 0;
+	handler(dev, id, ticks, ud);
```
is it possible that alarm is canceled and callback is still triggered? in that case `handler` may be null,

pabigot replied:
checked at call site; thanks.

galak commented:
DEV-REVIEW: Marking #17072 as 'TSC' with regards to the approach for how to have driver specific api's.  This is just an implementation of those ideas.

pabigot commented:
Excerpted [this comment on #17072](https://github.com/zephyrproject-rtos/zephyr/pull/17072#discussion_r311144317) by @MaureenHelm:
> Implementing the generic counter interface will make people think it's a counter. If you're worried about people thinking it's a counter, then why implement the generic counter interface at all?
>
> Similarly, if you think applications will use the device-specific functions exclusively, then isn't implementing the generic counter interface a waste of code space? I think we either need to revisit the RTC driver API or you should have a driver-specific API.

The description in #17072 was based on early work.  I believe in the final version the usefulness as a counter is more clear: in particular that you can set alarms based on a `time_t` value without having to mess with the driver-specific API, and you can read the current time using the standard API.

More generally I think an RTC API (which may or may not extend Counter) may be useful, but we need more examples of capability before we can figure out what it should look like.

pabigot commented:
All current review comments addressed.  Because I expect subsequent work to require extensions to civic time support in Zephyr I'm not going to push this for 2.0.0, as that would complicate the API changes.

@dbkinder Please check new commit documenting generic approach to API extension.

dbkinder commented on doc/reference/drivers/index.rst:
```diff
@@ -180,6 +180,60 @@ in the driver init function.
+standard API.  These devices combine subsystem operations with
+device-specific APIs, described in a device-specific header.
+
+A device-specific API defining typically looks like this:
```
change ``defining`` to ``definition``

dbkinder commented on samples/drivers/rtc_ds3231/README.rst:
```diff
@@ -0,0 +1,154 @@
+   .
+   DS3231 has not experienced an oscillator fault
+
+Next information about the device as a counter is displayed.  The
```
Change to ``Next, ``

dbkinder reviewed (changes requested):
A couple of typos, otherwise docs look good.

pabigot commented:
> A couple of typos, otherwise docs look good.

@dbkinder requested changes made.  Thanks.

pizi-nordic commented on drivers/counter/ds3231.c:
```diff
@@ -0,0 +1,1063 @@
+	}
+	data->rtc_base = decode_rtc(data);
+
+	printk("Time %u at %u (%d)\n", (u32_t)data->rtc_base,
```
This looks like some debug code left behind.

pabigot replied:
removed

nordic-krch replied:
is it?

pabigot replied:
Was probably lost during the restoration when I deleted the branch and github wouldn't allow the PR to be reopened.  Gone now.

dbkinder reviewed (approved):
+1 for docs, thanks

jhedberg commented:
Any update on this PR? Could the assigned reviewers (especially the codeowner ones) take a look please?

jhedberg commented:
Is this PR waiting for something (other than more reviews) before it's ready for merging? @pabigot looks like you'll need to fix a CODEOWNERS conflict either way.

pabigot commented:
This is only waiting for more reviews, but it's been waiting so long it does need to be rebased and retested.  There's more interest now in supporting civil time in Zephyr, so I'll try to do that early in the 2.3 development cycle.

pabigot commented:
This has been updated again.  Please take a look; I'd like to get it into 2.3.

nordic-krch commented on samples/drivers/rtc_ds3231/prj.conf:
```diff
@@ -0,0 +1,19 @@
+CONFIG_PRINTK=y
+
+# The DS3231 is an I2C device...
+CONFIG_I2C=y
```
it's already selected by `COUNTER_DS3231`

nordic-krch commented on include/drivers/rtc/ds3231.h:
```diff
@@ -0,0 +1,640 @@
+	rtc_ds3231_api_set set;
+	rtc_ds3231_api_check_alarms check_alarms;
+};
+
```
would be good to add build time assert which ensures that `counter_api` is the first item in the api struct.

pabigot replied:
We don't generally do that in similar cases, like gpio which has required common prefixes.  It'd be pretty obvious if it was wrong.

nordic-krch reviewed (commented):
it didn't go very deep and have no way of testings. Left few some comments.

GeorgeGkinis commented:
Nice work!

I see this PR has been waiting for 8 months.

Any chance it will be available for 2.3?

pabigot commented:
> Any chance it will be available for 2.3?

Probably not, but we'll try.  The latest blocker is support for userspace, which I understand to be a requirement for all new API.  So this is going back into DNM mode until that's resolved; some background is in #24539.

pabigot committed:
tests: drivers: counter: add rtc_ds3231_api

Cloned from counter_basic_api with modifications based on DS3231
limitations:
* Only one device tested per board;
* Counter cannot be stopped or started;
* Alarms are serviced by worker thread, so are not invoked from ISR
  and require that test thread yield to allow processing (no
  k_busy_wait());
* Multiple Alarms test is disabled as documented in test.

Additional tests were added for DS3231-specific API.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
This has been rebased and updated with proper locking and usermode support for the extended sync functions and the underlying counter API.  I would like to see something like the `rtc_ds3231_{req,get}_syncpoint()` API in whatever we use for RTC extensions.

Please provide feedback, particularly @jdascenzio @henrikbrixandersen  @sslupsky  who have expressed interest in RTC.  I would like to get this in to 2.3, rather than attempt to beat @nordic-krch 's record of 367 days from PR open to merge.

jfischer-phytec-iot commented on drivers/counter/ds3231.c:
```diff
@@ -0,0 +1,1368 @@
+
+	/* Flags recording requests for ISW monitoring. */
+	u8_t isw_mon_req;
+#define ISW_MON_REQ_Alarm 0x01
```
That is unusual.

pabigot replied:
You mean mixed case?  I suppose, though it's standard for some roles (e.g. CMSIS-style bitfield values), and is also present internally Zephyr's time_utils.h where the units suffix is (correctly) in lower case.

If it really bothers people I'll uglify it.

jfischer-phytec-iot replied:
I meant to define a macro in middle of a struct.

carlescufi commented:
@henrikbrixandersen and @jdascenzio could you please review, especially in the context of the work for #23526

pabigot commented:
Rebased and all public symbols renamed from `rtc_ds3231_` to `maxim_ds3231_` in anticipation of possible dedicated RTC API.

jfischer-phytec-iot commented on include/drivers/rtc/maxim_ds3231.h:
```diff
@@ -0,0 +1,602 @@
+ * @param user_data the corresponding parameter from
+ * maxim_ds3231_alarm::user_data.
+ */
+typedef void (*maxim_ds3231_alarm_callback_handler_t)(struct device *dev,
```
`maxim_ds3231_alarm_callback_handler_t` is to long, please use shorter names, here `ds3231_alarm_cb_handler` is fine.

Overall just ds3231/DS3231 instead of maxim_ds3231/MAXIM_DS3231 would be sufficient.

Why not generally abbreviate `callback` with `cb` if the prefix give the only feasible interpretation like `ds3231_notify_cb`?

pabigot replied:
While I started with `ds3231_` as the prefix, and that has prior art in tree for (e.g.) `ccs811_`, Zephyr has not carefully managed namespaces and I don't want to assume that there will never be conflicts between device model numbers across vendors.  Standardizing on the compatible name makes sense.  The negative impact is small, only affecting those who are using the custom API.

This proposed policy is something we should discuss more widely.  So: #24978

`callback` is unambiguous while `cb` is only clear in context with interpretation.  IMO the biggest problem with long-but-clear names is a consequence of 8-space indentation with a 78 character line limit.  I have no reason to believe the shorter version ever directly improves comprehension, and some reason to believe it makes it worse (increased cognitive load from requiring an interpretation at all).

I'm seeing 44 `foo_callback_t` typedefs, and 96 `foo_cb_t` typedefs, in Zephyr today.

Not to say I won't change it if people insist, but that's why I picked the names that I did.

henrikbrixandersen reviewed (approved):
I have mainly reviewed the DS3231 specific API extensions. Looks good!

I think we should get this merged. Once we have more RTC drivers in-tree, we can look at extending the counter API in a generic fashion or even defining a new, RTC specific API.

pabigot commented on samples/drivers/counter/maxim_ds3231/sample.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  sample.basic.maxim_ds3231:
+    build_only: true
+    platform_whitelist: efr32mg_sltb004a frdm_k64f nrf51_pca10028 nucleo_l476rg particle_xenon
```
The ChronoDot isn't physically compatible with Arduino headers, and at least two boards that are supported don't have them either.  So I'm reluctant to do that.

sslupsky replied:
> I think we should get this merged. Once we have more RTC drivers in-tree, we can look at extending the counter API in a generic fashion or even defining a new, RTC specific API.

There are a few drivers in the tree already.  Many of the existing SOC's have RTC devices on-chip.  Zephyr is using them for the system timer (as opposed to a counter).  So, in this regard, I think it is quite important to advance the discussion of an RTC api before it becomes a terrible mess to fix in the future.

pabigot commented on doc/reference/peripherals/index.rst:
```diff
@@ -27,6 +27,7 @@ Peripherals
    pwm.rst
    ps2.rst
    peci.rst
+   rtc.rst
```
The documentation for the driver-specific extensions ends up here.  It does state that it's a placeholder.

carlescufi commented:
@nordic-krch would you mind approving this if you have no objections?

carlescufi commented:
@jfischer-phytec-iot @sslupsky @nordic-krch please approve if you are fine with this PR as-is.

carlescufi commented:
Merging this since it's been open for far too long. If you have additional comments please voice them to @pabigot and they can go in as bug fixes during the 2.3 stabilization period.

cederom commented:
This is a nice starting point thank you but also requires external chip. I have created a feature request here: https://github.com/zephyrproject-rtos/zephyr/issues/35333 :-)
